File: backend/collectors/coingecko.py
"""
CoinGecko data collector - Market Cap, FDV, Supply
"""
import httpx
from typing import Dict, Any, Optional, List

COINGECKO_API = "https://api.coingecko.com/api/v3"


class CoinGeckoCollector:
    def __init__(self):
        self.client = httpx.AsyncClient(timeout=15.0)

    async def close(self):
        await self.client.aclose()

    async def fetch_coin_data(self, coin_id: str) -> Dict[str, Any]:
        """Fetch single coin data"""
        try:
            response = await self.client.get(
                f"{COINGECKO_API}/coins/{coin_id}",
                params={
                    "localization": "false",
                    "tickers": "false",
                    "market_data": "true",
                    "community_data": "false",
                    "developer_data": "false",
                    "sparkline": "false",
                },
            )
            data = response.json()
            market = data.get("market_data", {})

            return {
                "market_cap": market.get("market_cap", {}).get("usd"),
                "fully_diluted_valuation": market.get("fully_diluted_valuation", {}).get("usd"),
                "circulating_supply": market.get("circulating_supply"),
                "total_supply": market.get("total_supply"),
                "current_price": market.get("current_price", {}).get("usd"),
            }
        except Exception as e:
            print(f"CoinGecko error for {coin_id}: {e}")
            return {}

    async def fetch_markets(self, coin_ids: List[str]) -> Dict[str, Dict[str, Any]]:
        """Fetch multiple coins at once (more efficient)"""
        if not coin_ids:
            return {}

        result = {}
        try:
            response = await self.client.get(
                f"{COINGECKO_API}/coins/markets",
                params={
                    "vs_currency": "usd",
                    "ids": ",".join(coin_ids),
                    "order": "market_cap_desc",
                    "per_page": 250,
                    "page": 1,
                    "sparkline": "false",
                },
            )
            data = response.json()

            if isinstance(data, list):
                for coin in data:
                    result[coin["id"]] = {
                        "market_cap": coin.get("market_cap"),
                        "fully_diluted_valuation": coin.get("fully_diluted_valuation"),
                        "circulating_supply": coin.get("circulating_supply"),
                        "total_supply": coin.get("total_supply"),
                        "current_price": coin.get("current_price"),
                        "price_change_24h": coin.get("price_change_percentage_24h"),
                        "volume_24h": coin.get("total_volume"),
                    }
        except Exception as e:
            print(f"CoinGecko markets error: {e}")

        return result

File: src/lib/collectors/coingecko.ts
// CoinGecko data collector

import axios from "axios";

const COINGECKO_API = "https://api.coingecko.com/api/v3";

export interface CoinGeckoMetrics {
  marketCap: number | null;
  fullyDilutedValuation: number | null;
  circulatingSupply: number | null;
  totalSupply: number | null;
  currentPrice: number | null;
  priceChange24h: number | null;
  volume24h: number | null;
}

export interface CoinGeckoMarketData {
  id: string;
  symbol: string;
  name: string;
  current_price: number;
  market_cap: number;
  fully_diluted_valuation: number | null;
  circulating_supply: number;
  total_supply: number | null;
  price_change_percentage_24h: number;
  total_volume: number;
}

/**
 * Fetch coin data from CoinGecko
 */
export async function fetchCoinGeckoData(coinId: string): Promise<CoinGeckoMetrics> {
  try {
    const response = await axios.get(`${COINGECKO_API}/coins/${coinId}`, {
      params: {
        localization: false,
        tickers: false,
        market_data: true,
        community_data: false,
        developer_data: false,
        sparkline: false,
      },
      timeout: 15000,
    });

    const marketData = response.data.market_data;

    return {
      marketCap: marketData?.market_cap?.usd || null,
      fullyDilutedValuation: marketData?.fully_diluted_valuation?.usd || null,
      circulatingSupply: marketData?.circulating_supply || null,
      totalSupply: marketData?.total_supply || null,
      currentPrice: marketData?.current_price?.usd || null,
      priceChange24h: marketData?.price_change_percentage_24h || null,
      volume24h: marketData?.total_volume?.usd || null,
    };
  } catch (error) {
    console.error(`Error fetching CoinGecko data for ${coinId}:`, error);
    return {
      marketCap: null,
      fullyDilutedValuation: null,
      circulatingSupply: null,
      totalSupply: null,
      currentPrice: null,
      priceChange24h: null,
      volume24h: null,
    };
  }
}

/**
 * Fetch multiple coins data from CoinGecko markets endpoint
 */
export async function fetchCoinGeckoMarkets(
  coinIds: string[]
): Promise<Map<string, CoinGeckoMetrics>> {
  const result = new Map<string, CoinGeckoMetrics>();

  if (coinIds.length === 0) return result;

  try {
    const response = await axios.get<CoinGeckoMarketData[]>(`${COINGECKO_API}/coins/markets`, {
      params: {
        vs_currency: "usd",
        ids: coinIds.join(","),
        order: "market_cap_desc",
        per_page: 250,
        page: 1,
        sparkline: false,
        price_change_percentage: "24h",
      },
      timeout: 15000,
    });

    for (const coin of response.data) {
      result.set(coin.id, {
        marketCap: coin.market_cap,
        fullyDilutedValuation: coin.fully_diluted_valuation,
        circulatingSupply: coin.circulating_supply,
        totalSupply: coin.total_supply,
        currentPrice: coin.current_price,
        priceChange24h: coin.price_change_percentage_24h,
        volume24h: coin.total_volume,
      });
    }

    return result;
  } catch (error) {
    console.error("CoinGecko API error:", error);
    return result;
  }
}

/**
 * Search for a coin on CoinGecko by symbol
 */
export async function searchCoinGecko(query: string): Promise<{ id: string; symbol: string; name: string }[]> {
  try {
    const response = await axios.get(`${COINGECKO_API}/search`, {
      params: { query },
      timeout: 10000,
    });

    return response.data.coins.slice(0, 10).map((coin: { id: string; symbol: string; name: string }) => ({
      id: coin.id,
      symbol: coin.symbol,
      name: coin.name,
    }));
  } catch (error) {
    console.error("Error searching CoinGecko:", error);
    return [];
  }
}

/**
 * Check if CoinGecko API is accessible
 */
export async function checkCoinGeckoHealth(): Promise<boolean> {
  try {
    const response = await axios.get(`${COINGECKO_API}/ping`, { timeout: 5000 });
    return response.data.gecko_says === "(V3) To the Moon!";
  } catch {
    return false;
  }
}
